File: go/rz/riskscore.go
package rz

// Prevention layer: deterministic, named-factor risk scoring.
// Transparent (named, inspectable factors with an audit trail), merchant-
// portable (pure function over your own data), compliance-native. No black
// box — every factor is readable by a judge.

type RiskDecision string

const (
	RiskLow    RiskDecision = "low"
	RiskMedium RiskDecision = "medium"
	RiskHigh   RiskDecision = "high"
)

const (
	RISK_THRESHOLD_HIGH   = 60
	RISK_THRESHOLD_MEDIUM = 35
)

type RiskFactor struct {
	Key          string
	Label        string
	Value        float64
	Contribution float64
}

type CustomerHistory struct {
	PriorDeclineCount90d    int
	MandateAgeDays          int
	DaysSinceLastDecline    int
	PriorChargebackCount90d int
	CardExpiresWithinDays   *int
	RenewalCount            int
	AmountInRupees          float64
}

type RiskInput struct {
	History CustomerHistory
}

type RiskScore struct {
	Score    float64
	Decision RiskDecision
	Factors  []RiskFactor
}

func clampTo100(n float64) float64 {
	if n < 0 {
		return 0
	}
	if n > 100 {
		return 100
	}
	return n
}

// computeRiskFactors returns each small, named, inspectable contribution.
func computeRiskFactors(h CustomerHistory) []RiskFactor {
	var factors []RiskFactor

	declineContribution := min3float(float64(h.PriorDeclineCount90d), 3) * 20
	factors = append(factors, RiskFactor{
		Key:          "prior_decline_count_90d",
		Label:        "Prior declines (90d)",
		Value:        float64(h.PriorDeclineCount90d),
		Contribution: declineContribution,
	})

	mandateAge := h.MandateAgeDays
	mandateContribution := float64(0)
	if mandateAge < 30 {
		mandateContribution = 15
	} else if mandateAge < 90 {
		mandateContribution = 5
	}
	factors = append(factors, RiskFactor{
		Key:          "mandate_age_days",
		Label:        "Mandate age (days)",
		Value:        float64(mandateAge),
		Contribution: mandateContribution,
	})

	recency := h.DaysSinceLastDecline
	recencyContribution := float64(0)
	if recency < 3 {
		recencyContribution = 20
	} else if recency < 14 {
		recencyContribution = 12
	} else if recency < 30 {
		recencyContribution = 5
	}
	factors = append(factors, RiskFactor{
		Key:          "days_since_last_decline",
		Label:        "Days since last decline",
		Value:        float64(recency),
		Contribution: recencyContribution,
	})

	chargebackContribution := min3float(float64(h.PriorChargebackCount90d), 3) * 25
	factors = append(factors, RiskFactor{
		Key:          "prior_chargeback_count_90d",
		Label:        "Prior chargebacks (90d)",
		Value:        float64(h.PriorChargebackCount90d),
		Contribution: chargebackContribution,
	})

	if h.CardExpiresWithinDays != nil && *h.CardExpiresWithinDays <= 60 {
		factors = append(factors, RiskFactor{
			Key:          "card_expiry_within_60d",
			Label:        "Card expires within 60d",
			Value:        float64(*h.CardExpiresWithinDays),
			Contribution: 18,
		})
	}

	return factors
}

func min3float(n, cap float64) float64 {
	if n < 0 {
		return 0
	}
	if n > cap {
		return cap
	}
	return n
}

// RiskScore computes a 0..100 deterministic risk score with named factors.
// The *decision* uses the un-clamped sum, mirroring the TS threshold logic;
// the returned score clamps to 0..100. Factors with contribution 0 are dropped.
func riskScore(input RiskInput) RiskScore {
	factors := computeRiskFactors(input.History)
	sum := float64(0)
	for _, f := range factors {
		sum += f.Contribution
	}
	decision := RiskLow
	if sum >= RISK_THRESHOLD_HIGH {
		decision = RiskHigh
	} else if sum >= RISK_THRESHOLD_MEDIUM {
		decision = RiskMedium
	}

	// topFactors: sorted desc by contribution, then filtered > 0.
	top := make([]RiskFactor, 0, len(factors))
	for _, f := range factors {
		if f.Contribution > 0 {
			top = append(top, f)
		}
	}
	// insertion sort desc by contribution (stable)
	for i := 1; i < len(top); i++ {
		for j := i; j > 0 && top[j].Contribution > top[j-1].Contribution; j-- {
			top[j], top[j-1] = top[j-1], top[j]
		}
	}

	return RiskScore{Score: clampTo100(sum), Decision: decision, Factors: top}
}

// IsHighRisk reports whether the risk decision is 'high'.
func IsHighRisk(r RiskScore) bool { return r.Decision == RiskHigh }

// RiskScorePublic mirrors TS's exported riskScore signature.
func RiskScorePublic(history CustomerHistory) RiskScore {
	return riskScore(RiskInput{History: history})
}

File: src/risk/riskScore.ts
// Prevention layer: deterministic, named-factor risk scoring.
// Mirrors the idea behind Razorpay's Vulcan (pattern-based payment
// intelligence) but is: transparent (named, inspectable factors with an audit
// trail), merchant-portable (pure function over your own data), and
// compliance-native. No black box — a judge can read every factor.

export type RiskDecision = 'low' | 'medium' | 'high';

export interface RiskFactor {
  key: string; // stable, nameable factor id
  label: string;
  value: number; // raw value
  contribution: number; // weighted points added to score
}

export interface RiskInput {
  // Customer payment/transaction history (derivable from your own data).
  history: CustomerHistory;
  // Optional current-event signal (error code, payment method, etc.).
}

export interface CustomerHistory {
  priorDeclineCount90d: number; // # declines in the last 90 days
  mandateAgeDays: number; // age of the mandate/subscription in days
  daysSinceLastDecline: number; // days since the most recent decline
  priorChargebackCount90d: number; // chargebacks in last 90 days
  cardExpiresWithinDays?: number; // days until card expiry (null = unknown)
  renewalCount?: number; // how many renewals already succeeded
  amountInRupees?: number; // current attempt amount
}

export interface RiskScore {
  score: number; // 0..100
  decision: RiskDecision;
  factors: RiskFactor[];
}

export const RISK_THRESHOLD_HIGH = 60;
export const RISK_THRESHOLD_MEDIUM = 35;

const clampTo100 = (n: number): number => Math.max(0, Math.min(100, n));

// Each factor is a small, named, inspectable contribution.
export function computeRiskFactors(h: CustomerHistory): RiskFactor[] {
  const factors: RiskFactor[] = [];

  // (a) Prior declines in 90d: each decline raises risk (capped at 3 for max).
  const declineContribution = Math.min(3, Math.max(0, h.priorDeclineCount90d)) * 20;
  factors.push({
    key: 'prior_decline_count_90d',
    label: 'Prior declines (90d)',
    value: h.priorDeclineCount90d,
    contribution: declineContribution,
  });

  // (b) Mandate age: brand-new mandate (<30d) is higher risk for autopay.
  const mandateAge = h.mandateAgeDays;
  const mandateContribution = mandateAge < 30 ? 15 : mandateAge < 90 ? 5 : 0;
  factors.push({
    key: 'mandate_age_days',
    label: 'Mandate age (days)',
    value: mandateAge,
    contribution: mandateContribution,
  });

  // (c) Recency of last decline: more recent = elevated risk.
  const recency = h.daysSinceLastDecline;
  const recencyContribution =
    recency < 3 ? 20 : recency < 14 ? 12 : recency < 30 ? 5 : 0;
  factors.push({
    key: 'days_since_last_decline',
    label: 'Days since last decline',
    value: recency,
    contribution: recencyContribution,
  });

  // (d) Chargebacks: strong fraud/risk signal.
  const chargebackContribution = Math.min(3, Math.max(0, h.priorChargebackCount90d)) * 25;
  factors.push({
    key: 'prior_chargeback_count_90d',
    label: 'Prior chargebacks (90d)',
    value: h.priorChargebackCount90d,
    contribution: chargebackContribution,
  });

  // (e) Card expiring soon: high risk of card_expired bucket.
  if (h.cardExpiresWithinDays != null && h.cardExpiresWithinDays <= 60) {
    factors.push({
      key: 'card_expiry_within_60d',
      label: 'Card expires within 60d',
      value: h.cardExpiresWithinDays,
      contribution: 18,
    });
  }

  return factors;
}

export function riskScore(input: RiskInput): RiskScore {
  const factors = computeRiskFactors(input.history);
  const score = factors.reduce((s, f) => s + f.contribution, 0);
  const decision: RiskDecision =
    score >= RISK_THRESHOLD_HIGH ? 'high' : score >= RISK_THRESHOLD_MEDIUM ? 'medium' : 'low';
  const topFactors = [...factors].sort((a, b) => b.contribution - a.contribution).filter((f) => f.contribution > 0);
  return { score: clampTo100(score), decision, factors: topFactors };
}

// A named decision the recovery layer can act on preemptively.
export function isHighRisk(r: RiskScore): boolean {
  return r.decision === 'high';
}
